### PR TITLE
JobVerificationEngine DAG construction (no driver assigned partition IDs)

### PR DESCRIPTION
**Pull request by andrewlawhh**
Refactor JobVerificationEngine removing trust from driver (no jobID assigned by driver)

- Create graph based on input/output MACs.

### CODE REVIEW

File: src/enclave/Enclave/EnclaveContext.h [hunk 1]
```diff
@@ -7,56 +7,88 @@
...
+      if (this->input_log_macs.size() != c.input_log_macs.size()) {
+          return false;
+      }
+
+      // Check whether the input_log_macs themselves are the same
+      for (uint32_t i = 0; i < this->input_log_macs.size(); i++) {
+          if (this->input_log_macs[i] != c.input_log_macs[i]) {
+              return false;
+          }
+      }
```
Comment (wzheng) [inline src/enclave/Enclave/EnclaveContext.h:47]: I'm a bit confused. Is the `input_log_macs` a sequence of MACs of size `OE_HMAC_SIZE`? If so, does the equality condition also hold when the set of `this->input_log_macs` is the same as that of `c.input_log_macs`?

**Review by wzheng — COMMENTED** [review]
Partial review

File: src/enclave/Enclave/EnclaveContext.h [hunk 2]
```diff
@@ -7,56 +7,88 @@
...
+    // int pid;
     bool append_mac;
 
     // Map of job ID for partition
     std::unordered_map<int, int> pid_jobid;
 
 
     EnclaveContext() {
-      pid = -1;
+      // pid = -1;
```
Comment (chester-leung) [inline src/enclave/Enclave/EnclaveContext.h:90]: This too

File: src/enclave/Enclave/FlatbuffersWriters.cpp [hunk 3]
```diff
@@ -160,68 +160,142 @@ flatbuffers::Offset<tuix::EncryptedBlocks> RowWriter::finish_blocks(std::string
...
-      // char* untrusted_ecall_op_str = oe_host_strndup(le.ecall.c_str(), le.ecall.length());
-      auto past_log_entry_serialized = tuix::CreateLogEntry(enc_block_builder,
-          le.ecall,
-          le.snd_pid,
-          le.rcv_pid,
-          le.job_id);
-      past_log_entries_vector.push_back(past_log_entry_serialized);
+    // Serialize stored crumbs
+    std::vector<Crumb> crumbs = EnclaveContext::getInstance().get_crumbs();
+    // std::cout << "Num crumbs: " << crumbs.size() << std::endl;
```
Comment (chester-leung) [inline src/enclave/Enclave/FlatbuffersWriters.cpp:188]: Can you remove these print statements?

File: src/enclave/Enclave/FlatbuffersWriters.cpp [hunk 4]
```diff
@@ -160,68 +160,142 @@ flatbuffers::Offset<tuix::EncryptedBlocks> RowWriter::finish_blocks(std::string
...
-
-    // Copy the mac to untrusted memory
-    uint8_t* untrusted_mac = nullptr;
-    ocall_malloc(OE_HMAC_SIZE, &untrusted_mac);
-    std::unique_ptr<uint8_t, decltype(&ocall_free)> mac_ptr(untrusted_mac, &ocall_free);
-    memcpy(mac_ptr.get(), hmac, OE_HMAC_SIZE);
-    auto mac_offset = tuix::CreateMac(enc_block_builder, 
-        enc_block_builder.CreateVector(mac_ptr.get(), OE_HMAC_SIZE));
-    log_entry_chain_hash_vector.push_back(mac_offset);
+    // std::cout << "log entry num bytes: " << log_entry_num_bytes_to_mac << std::endl;
```
Comment (chester-leung) [inline src/enclave/Enclave/FlatbuffersWriters.cpp:256]: Remove these print statements too

File: src/enclave/Enclave/FlatbuffersWriters.cpp [hunk 5]
```diff
@@ -160,68 +160,142 @@ flatbuffers::Offset<tuix::EncryptedBlocks> RowWriter::finish_blocks(std::string
...
   } 
   auto log_entry_chain_serialized = tuix::CreateLogEntryChainDirect(enc_block_builder, 
-      &curr_log_entry_vector, &past_log_entries_vector, &num_past_log_entries);
+      &curr_log_entry_vector, &serialized_crumbs_vector, &num_crumbs_vector);
+
+  // Create dummy array that isn't default, so that we can modify it using Flatbuffers mutation later
+  uint8_t dummy_all_outputs_mac[OE_HMAC_SIZE] = {1};
+   
+  // // Copy the dummmy all_outputs_mac to untrusted memory
+  // uint8_t* untrusted_dummy_all_outputs_mac = nullptr;
```
Comment (chester-leung) [inline src/enclave/Enclave/FlatbuffersWriters.cpp:286]: delete these too

File: src/enclave/Enclave/IntegrityUtils.cpp [hunk 6]
```diff
@@ -124,78 +150,146 @@ void verify_log(const tuix::EncryptedBlocks *encrypted_blocks,
...
+          // crumb.all_outputs_mac, crumb.log_mac are of size OE_HMAC_SIZE
+          // crumb.input_macs is of size num_input_macs * OE_HMAC_SIZE
+          int num_bytes_in_crumb = 2 * sizeof(int) + 2 * OE_HMAC_SIZE + OE_HMAC_SIZE * crumbs[j].num_input_macs;
+          total_crumb_bytes += num_bytes_in_crumb;
+      }
+      // Below, we add sizeof(int) to include the num_past_entries entry that is part of LogEntryChain 
+      int total_bytes_to_mac = log_entry_num_bytes_to_mac + total_crumb_bytes + sizeof(int);
 
-      int num_bytes_to_mac = OE_HMAC_SIZE + 6 * sizeof(int) + num_past_entries * 4 * sizeof(int); 
+      // std::cout << "log entry num bytes: " << log_entry_num_bytes_to_mac << std::endl;
```
Comment (chester-leung) [inline src/enclave/Enclave/IntegrityUtils.cpp:170]: Delete these

**Review by chester-leung — COMMENTED** [review]
Can you also delete the commented out print statements / old code? I left comments at some places but not all.

**Review by chester-leung — APPROVED** [review]

**Pull request merged**